### DISCUSSION TITLE
Unable to save settings, tried multiple computers, tried different versions of LRR

### DISCUSSION THREAD

**Discussion by Albireo6972** [Difegue/LANraragi#534] [Q&A] [answered]
**LRR Version and OS**
Please tell which version you're running, as well as your OS and installation method. (Windows, Docker, Native)
Windows 10 Pro 21H1 build 19043.1288
LRR version Version 0.8.1, "Black Tie White Noise (Here Come Da Jazz Mix)"
Install method MSI install, Linux Subsystem installed, tried WSL 1 and WSL 2 within LRR settings.

**Bug Details**
Please detail the bug here. 

I am running into issues saving settings, I have tried installing Lanraragi on 3 separate devices but get the same error. When I save and check the log console I get this error:

Use of uninitialized value in subroutine entry at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Database.pm line 253.

And when I check the Redis Logs I get this error:

37:M 24 Oct 2021 23:21:46.239 * AOF rewrite child asks to stop sending diffs.
3106:C 24 Oct 2021 23:21:46.239 * Parent agreed to stop sending diffs. Finalizing AOF...
3106:C 24 Oct 2021 23:21:46.239 * Concatenating 0.00 MB of AOF diff received from parent.
3106:C 24 Oct 2021 23:21:46.282 * SYNC append only file rewrite performed
3106:C 24 Oct 2021 23:21:46.283 * AOF rewrite: 0 MB of memory used by copy-on-write
37:M 24 Oct 2021 23:21:46.295 * Background AOF rewrite terminated with success
37:M 24 Oct 2021 23:21:46.296 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
37:M 24 Oct 2021 23:21:46.297 # Error trying to rename the temporary AOF file temp-rewriteaof-bg-3106.aof into appendonly.aof: Permission denied
37:M 24 Oct 2021 23:21:46.400 * Background append only file rewriting started by pid 3107

I have uninstalled LRR, the Linux subsystem, restarted, installed, same thing. New database or old database and no change. I am unable to change any settings and save, it will give the message for success, but on exit and restart it will default back to initial settings. I have tried as both a normal click/run instance, and run as administrator instance, and sadly same issue. 

Any help on this matter would be appreciated.

**Matching Logs**
Post any logs you have encountered when reproducing the bug.  
If possible, switch the app to Debug Mode in order to have more logs.

Debug log of an attempt to save:

[LANraragi] [debug] Saving nofunmode with value 1
[LANraragi] [debug] Saving localprogress with value 1
Use of uninitialized value in subroutine entry at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Database.pm line 253.
[LANraragi] [debug] Saving thumbdir with value
[LANraragi] [debug] Saving apikey with value 
[LANraragi] [debug] Saving enableresize with value 0
[LANraragi] [debug] Saving readerquality with value 50
[LANraragi] [debug] Saving password with value 
[LANraragi] [debug] Saving enablecors with value 0
[LANraragi] [debug] Saving motd with value Welcome to this Library running LANraragi!
[LANraragi] [debug] Saving sizethreshold with value 1000
[LANraragi] [debug] Saving pagesize with value 100
[LANraragi] [debug] Saving tagrules with value -already uploaded;-russian;-chinese;-portuguese;-french;-spanish;-italian;-vietnamese;-german;-indonesian
Use of uninitialized value in subroutine entry at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Database.pm line 253.
[LANraragi] [debug] Saving dirname with value
[LANraragi] [debug] Saving tempmaxsize with value 500
[LANraragi] [debug] Saving enablepass with value 1
[LANraragi] [debug] Saving devmode with value 1
[LANraragi] [debug] Saving htmltitle with value LANraragi
[LANraragi] [debug] Saving tagruleson with value 1
[LANraragi] [debug] Saving computed tag rules : [["remove","already uploaded",""],["remove","russian",""],["remove","chinese",""],["remove","portuguese",""],["remove","french",""],["remove","spanish",""],["remove","italian",""],["remove","vietnamese",""],["remove","german",""],["remove","indonesian",""]]

**Screenshots**

![image](https://user-images.githubusercontent.com/46336030/138617052-0f077006-ffc0-4df7-aea1-57e57b7af551.png)

**Comment by Difegue** [comment]
Looks like a permission issue on an existing `appendonly.aof` file in your content folder.  
Try making a backup of that file and removing it from the folder to see if that changes anything - There should be no data loss due to the `database.rdb` backup, but you never know.

**Comment by Albireo6972** [comment]
Sadly deleting the appendonly.aof file didn't work either. I also just tried creating a new content folder, added a few files, set the directory, started LRR, and got the same thing. I am currently set to WSL 1 and ran it as both normal and administrator.

**Comment by Difegue** [comment]
Sadly I'm not really sure where the issue lies since it seems to be a hiccup between WSL and how redis saves its AOF files -- You can either rollback to 0.7.9 which does not have AOF, or remove [this line](https://github.com/Difegue/LANraragi/blob/c6b10ec6dc0d0939da06709709f8ff561d1e8385/lib/LANraragi.pm#L93) in the server's files to disable it.  

I'll run some tests to make sure AOF isn't busted on 0.8.1 + WSL later today.

**Comment by Albireo6972** [comment]
Reverting back to 0.7.9 has fixed the issue, I tried removing line # 93 but I would get the a 
different error which is as follows:

Use of uninitialized value in subroutine entry at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Database.pm line 253.

removing the block at 250-256 causes another error at line 119 of config.pm in /home/koyomi/lanraragi/script/../lib/LANraragi/Controller/Config.pm

And that is where I stopped, and decided it would be best not to mess with functions more than I should. 

 So it definitely seems to be an issue with WSL and redis saving the AOF files on my end. I have reverted back to 0.7.9 and I can save without any issues.

**Comment by Difegue** [comment, marked as answer]
I've ran some tests with the latest 0.8.1 on my Win11 machine and can confirm the AOF file and its RDB variant work just fine. 
Did you use the same content folder between all your installs, in a network folder for instance? That might be the cause of the problem. 

I've also tried removing only the line that enables AOF and didn't encounter the errors you mentioned:
![image](https://user-images.githubusercontent.com/8237712/139686813-1201df7d-f47a-4649-b160-13c819281128.png)  

Moving this to a Q&A discussion since it doesn't seem to stem from a software bug 👍

**Reply by Albireo6972** [reply]
Appreciate the info, I did use the same content folder for my installs. I have moved it to a new location, and installed the nightly, and it seems to have helped. I can save 8 out of 10 times, it still fails every once in a while, but I have figured out a way to at least get it to save and take. Thank you for the research, and I do apologize for sending you on a wild goose chase for my issue.
